src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs

```diff
@@ -0,0 +1,222 @@
+            }
+
+            this.finishedPolls = this.LoadPolls(FinishedPollsDbKey);
+
+            if (this.finishedPolls == null)
+            {
+                this.logger.LogDebug("No finished polls found in DB, initializing with empty collection.");
+
+                this.finishedPolls = new List<Poll>();
+                this.SavePolls(this.finishedPolls, FinishedPollsDbKey);
+            }
+        }
+
+        /// <summary>Schedules a vote for the next time when the block will be mined.</summary>
+        public void ScheduleVote(VotingData votingData)
+        {
+            if (!this.federationManager.IsFederationMember)
+            {
+                this.logger.LogTrace("(-)[NOT_FED_MEMBER]");
+                throw new InvalidOperationException("Not a federation member!");
+            }
+
+            lock (this.locker)
+            {
+                this.scheduledVotingData.Add(votingData);
+            }
+        }
+
+        /// <summary>Provides a copy of scheduled voting data.</summary>
+        public List<VotingData> GetScheduledVotes()
+        {
+            lock (this.locker)
+            {
+                return new List<VotingData>(this.scheduledVotingData);
+            }
+        }
+
+        /// <summary>Provides scheduled voting data and removes all items that were provided.</summary>
+        /// <remarks>Used by miner.</remarks>
+        public List<VotingData> GetAndCleanScheduledVotes()
+        {
+            lock (this.locker)
+            {
+                List<VotingData> votingData = this.scheduledVotingData;
+
+                this.scheduledVotingData = new List<VotingData>();
+
+                return votingData;
+            }
+        }
+
+        /// <summary>Saves collection of polls to the database under provided key.</summary>
+        private void SavePolls(List<Poll> polls, string key)
+        {
+            this.keyValueRepo.SaveValueJson(key, polls);
+        }
+
+        /// <summary>Loads collection of polls from the database using provided key.</summary>
+        private List<Poll> LoadPolls(string key)
+        {
+            List<Poll> polls = this.keyValueRepo.LoadValueJson<List<Poll>>(key);
+            return polls;
+        }
+
+        /// <summary>Provides a collection of polls that are currently active.</summary>
+        public List<Poll> GetPendingPolls()
+        {
+            lock (this.locker)
+            {
+                return new List<Poll>(this.pendingPolls);
+            }
+        }
+
+        /// <summary>Provides a collection of polls that are already finished and their results applied.</summary>
+        public List<Poll> GetFinishedPolls()
+        {
+            lock (this.locker)
+            {
+                return new List<Poll>(this.finishedPolls);
+            }
+        }
+
+        // TODO subscribe to signals
+        // TODO more logs and log polls
+        private void onBlockConnected(ChainedHeaderBlock chBlock)
```

Comment:
Come to think about it, I am not sure the voting db should be disconnected from consensus, it should probably be a full validation rule  that will update some db and that db will also rewind together with the engine rewind.

Reply:
As discussed in PM this is what happens- on block connect and disconnect are called under a reorg lock after block was connected and rewind happened. So due to the lack of other place to do this- this is the only place to add this logic.